core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp

```diff
@@ -217,64 +216,70 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
         cgh.parallel_for(
             sycl::nd_range<1>(n_wgroups * wgroup_size, wgroup_size),
             [=](sycl::nd_item<1> item) {
-              const auto local_id  = item.get_local_linear_id();
-              const auto global_id = item.get_global_linear_id();
+              const auto local_id = item.get_local_linear_id();
+              const auto global_id =
+                  wgroup_size * item.get_group_linear_id() * values_per_thread +
+                  local_id;
 
               // Initialize local memory
+              const typename Policy::index_type upper_bound =
+                  (global_id + values_per_thread * wgroup_size) < size
+                      ? global_id + values_per_thread * wgroup_size
+                      : size;
```

Comment:
`std::min` might be more readable than this ternary op

Reply:
Yes, I wasn't quite sure if that's supported on the device. I can check.